### PR TITLE
Cleaning: Clean code base a bit & swerve as code review assignment

### PR DESCRIPTION
**Pull request by SusanC3**

### CODE REVIEW

File: src/main/cpp/Robot.cpp [hunk 1]
```diff
@@ -27,25 +31,29 @@ void Robot::TeleopInit() {
 
 void Robot::TeleopPeriodic() {
 
-  double dx = ljoy.GetRawAxis(1);
```
Comment (bli-school-account) [inline src/main/cpp/Robot.cpp]: what is the difference between d and v here? why did you change it?

File: src/main/cpp/Robot.cpp [hunk 2]
```diff
@@ -4,20 +4,24 @@
...
 
 void Robot::AutonomousInit() {
-  swerveDrive_->initializeAutoTraj(SwerveConstants::twoBallPath); //todo would be done with auto chooser
+  swerveDrive_->initializeAutoTraj(SwerveConstants::testPath); //todo would be done with auto chooser depending on auto mode
 }
 
+//commented out for testing purposes
 void Robot::AutonomousPeriodic() {
   // swerveDrive_->setState(SwerveDrive::State::PATH_FOLLOW); //todo would be moved into auto executor
-  // swerveDrive_->Periodic(0); //dummy turret value of 0 for testing purposes
```
Comment (bli-school-account) [inline src/main/cpp/Robot.cpp:20]: does this function have 1 parameter or 4 parameters?

File: src/main/cpp/Robot.cpp [hunk 3]
```diff
@@ -4,20 +4,24 @@
...
   swerveDrive_ = new SwerveDrive(navx_, limelight_);
 }
 
-void Robot::RobotPeriodic() {}
+void Robot::RobotPeriodic() {
+  limelight_.lightOn(false);
+}
 
 void Robot::AutonomousInit() {
-  swerveDrive_->initializeAutoTraj(SwerveConstants::twoBallPath); //todo would be done with auto chooser
```
Comment (bli-school-account) [inline src/main/cpp/Robot.cpp]: why does the purpose of this function change?

File: src/main/cpp/Robot.cpp [hunk 4]
```diff
@@ -27,25 +31,29 @@ void Robot::TeleopInit() {
...
-  dtheta = abs(dtheta) < 0.05 ? 0.0: dtheta;
-
-  joy_val_to_mps(dx);
-  joy_val_to_mps(dy);
-  joy_rot_to_rps(dtheta);
-
+  //get joystick input
+  double vx = ljoy.GetRawAxis(1);
+  double vy = ljoy.GetRawAxis(0);
+  double vtheta = rjoy.GetX();
```
Comment (AlwaysAcing123) [inline src/main/cpp/Robot.cpp:37]: What is ljoy and rjoy, and what is the difference between .GetRawAxis(1) and GetX()?

File: src/main/include/SwerveModule.h [hunk 5]
```diff
@@ -25,12 +25,13 @@ class SwerveModule
         frc::SwerveModuleState getState();
         frc::SwerveModuleState getOptState(frc::SwerveModuleState state);
         
+        units::meters_per_second_t getVelocityMPS();
         double getVelocity();
         double getYaw();
 
         void setAngMotorVoltage(double voltage);
```
Comment (AlwaysAcing123) [inline src/main/include/SwerveModule.h]: What does this function do? (setAngMotorVoltage())

File: src/main/cpp/Robot.cpp [hunk 6]
```diff
@@ -27,25 +31,29 @@ void Robot::TeleopInit() {
...
-  dx = abs(dx) < 0.1 ? 0.0: dx; 
-  dy = abs(dy) < 0.1 ? 0.0: dy;
-  dtheta = abs(dtheta) < 0.05 ? 0.0: dtheta;
-
-  joy_val_to_mps(dx);
-  joy_val_to_mps(dy);
-  joy_rot_to_rps(dtheta);
-
+  //get joystick input
+  double vx = ljoy.GetRawAxis(1);
```
Comment (bli-school-account) [inline src/main/cpp/Robot.cpp:35]: why is this called "raw" and what does it do?

File: src/main/cpp/Robot.cpp [hunk 7]
```diff
@@ -27,25 +31,29 @@ void Robot::TeleopInit() {
...
+  double vy = ljoy.GetRawAxis(0);
+  double vtheta = rjoy.GetX();
+  //apply deadband
+  vx = abs(vx) < 0.1 ? 0.0: vx;
```
Comment (bli-school-account) [inline src/main/cpp/Robot.cpp]: what is the question mark for in this code?

**Review by bli-school-account — COMMENTED** [review]
comment

**Review by linc-mvla — APPROVED** [review]
Looks all good to me

File: src/main/cpp/Robot.cpp (hunk 1 above)
Comment (SusanC3) [reply src/main/cpp/Robot.cpp]: The d stood for derivative, so like the derivative of x would be the velocity in the x direction. But since many people haven't taken calculus, I figured it'd be more clear to have a v instead to stand for velocity.

File: src/main/cpp/Robot.cpp (hunk 3 above)
Comment (SusanC3) [reply src/main/cpp/Robot.cpp]: The parameter passed into this function indicates with path to initialize. When I first wrote this, I made a sample path for a two ball auto. I changed it to a test path (which just goes straight forward 3 meters) to prepare for testing.

File: src/main/cpp/Robot.cpp (hunk 4 above)
Comment (SusanC3) [reply src/main/cpp/Robot.cpp:37]: They are basically the same, the raw axis function just allows the axis to be specified by parameter instead of a different function. I agree the use of redundant functions is unclear; I'll check which ljoy axis is which next time I'm in shop and I'll change this.

File: src/main/include/SwerveModule.h (hunk 5 above)
Comment (SusanC3) [reply src/main/include/SwerveModule.h]: The `getVelocityMPS()` function returns the linear velocity of the swerve module's driving wheel in meters per second. This is calculated given the angular velocity of the wheel and the wheel's radius. 

The `setAngMotorVoltage()` function is just a way to pass in a voltage value output to the swerve module's angle motor, the one that controls the driving wheel's orientation. There are different output modes that can be used when setting a motor (for example, percent output), so I specifically included the word voltage to not be unclear.

File: src/main/cpp/Robot.cpp (hunk 6 above)
Comment (SusanC3) [reply src/main/cpp/Robot.cpp:35]: It's called raw because for this function you pass in the "raw" axis number instead of calling the `getX()` and `getY()` axis getter functions. Mentioned this in an earlier comment, but since the function redundancy is confusing, next time I'm in shop I'll change this to be `getX()` or `getY()`, whichever one it is.

File: src/main/cpp/Robot.cpp (hunk 7 above)
Comment (SusanC3) [reply src/main/cpp/Robot.cpp]: This is a shorthand way to do an if statement, otherwise known as the ternary operator. This line of code is equivalent to the if statement 
`if (abs(vx) < 0.1) {`
`        vx = 0.0;`
`} else {`
`       vx = vx;`
`}`
You can read more about inline if statements/ternary operators here: [https://www.w3schools.com/cpp/cpp_conditions_shorthand.asp](https://www.w3schools.com/cpp/cpp_conditions_shorthand.asp)

File: src/main/cpp/SwerveDrive.cpp [hunk 8]
```diff
@@ -54,6 +56,7 @@ void SwerveDrive::updateOdometry(frc::Rotation2d robotAngle, frc::Pose2d robotPo
 **/
 void SwerveDrive::initializeAutoTraj(std::string filePath) {
     ramseteCommand_ = setupRamsete(filePath); 
+    ramseteCommand_->Initialize();
```
Comment (bli-school-account) [inline src/main/cpp/SwerveDrive.cpp]: how does this arrow -> work in c++?

File: src/main/cpp/SwerveDrive.cpp [hunk 9]
```diff
@@ -147,24 +150,44 @@ void SwerveDrive::drive(units::meters_per_second_t dx, units::meters_per_second_
         -GeneralConstants::MAX_VOLTAGE, GeneralConstants::MAX_VOLTAGE) );
 
     flModule_.setSpeedMotor( 0.2*std::clamp(fl_opt.speed.value(), -1.0, 1.0) );
+    // frc::SmartDashboard::PutNumber("fl speed error", fl_opt.speed.value() - flModule_.getVelocityMPS().value());
```
Comment (bli-school-account) [inline src/main/cpp/SwerveDrive.cpp]: what is the opt thing here (fl_opt, fr_opt, etc) and what is the yaw thing on line 135(navx_yaw) and what do these two things do?

**Review by bli-school-account — COMMENTED** [review]
comment

File: src/main/cpp/SwerveDrive.cpp [hunk 10]
```diff
@@ -78,12 +81,12 @@ frc::DifferentialDriveWheelSpeeds SwerveDrive::getDifferentialWheelSpeeds() {
...
+ * @param vx the desired robot velocity in the x direction
+ * @param vy the desired robot velocity in the y direction
+ * @param vtheta the desired robot rotational velocity
  * @param turretAngle the turret's angle on the robot, used to help calculate odometry
 **/
-void SwerveDrive::Periodic(units::meters_per_second_t dx, units::meters_per_second_t dy, units::radians_per_second_t dtheta, double turretAngle) {
+void SwerveDrive::Periodic(units::meters_per_second_t vx, units::meters_per_second_t vy, units::radians_per_second_t vtheta, double turretAngle) {
 
     //update odometry no matter the state
     odometry_->Update(units::degree_t{navx_->GetYaw()}, getRealModuleStates());
```
Comment (bli-school-account) [inline src/main/cpp/SwerveDrive.cpp]: what is the getRealModuleStates() thing here and what does it do

File: src/main/cpp/SwerveDrive.cpp [hunk 11]
```diff
@@ -104,7 +107,7 @@ void SwerveDrive::Periodic(units::meters_per_second_t dx, units::meters_per_seco
     //takes appropriate action based on current state
     switch(state_) {
         case DRIVE:
-            drive(dx, dy, dtheta, turretAngle);
+            drive(vx, vy, vtheta, turretAngle);
             break;
         case PATH_FOLLOW:
             ramseteCommand_->Execute();
```
Comment (bli-school-account) [inline src/main/cpp/SwerveDrive.cpp]: what is this ramsete thing? I googled it on WPILib but did not fully understand what it did

File: src/main/cpp/SwerveDrive.cpp (hunk 8 above)
Comment (SusanC3) [reply src/main/cpp/SwerveDrive.cpp]: The arrow (->) and dot (.) operator are basically the same, they reference a public member (variable, function, etc) of an instance of a class. You use the arrow when the instance is a pointer, and a dot when the instance is not a pointer.  Here the `ramseteCommand_` object is a pointer, so the arrow is used.

File: src/main/cpp/SwerveDrive.cpp (hunk 9 above)
Comment (SusanC3) [reply src/main/cpp/SwerveDrive.cpp]: The "opt" or optimized module states basically transforms the initial calculated goal state into a state that is "easier" for the module to achieve. For example, say the wheel is currently at an angle of 0°, and the calculated goal state wants the wheel to be at 180°. Since 0° and 180° are really the same position, just with the wheel spinning the opposite direction, it's easier for the wheel to switch directions and spin backwards instead of rotation 180°. The optimize function would take in the state saying the module should rotate 180° and return a state saying the module should not rotate at all and instead reverse its direction.

File: src/main/cpp/SwerveDrive.cpp (hunk 10 above)
Comment (SusanC3) [reply src/main/cpp/SwerveDrive.cpp]: `getRealModuleStates()` (defined [here](https://github.com/Eaglestrike/2023-RobotCode/blob/master/src/main/cpp/SwerveDrive.cpp#L59L66)) just returns the current state of each module (so the measured speed and angle of the wheel) combined into an array of 4 module states. I'm basically just restating what the comment says; does that make sense? Or should I update the comment?

Odometry works by recording the speed and angle of the wheels at each time step and "integrating" them to determine the robot's position. When the robot is given information about its module states, it can figure out its velocity (speed and direction). It can then use the `change in distance = velocity*time` equation, substituting in the current velocity and the time period between each cycle of the code being run. It then adds that change in position to last cycle's position, and in that manner the robot calculates its position on the field.

File: src/main/cpp/SwerveDrive.cpp (hunk 11 above)
Comment (SusanC3) [reply src/main/cpp/SwerveDrive.cpp]: The `ramseteCommand_` object contains

- A trajectory for the robot to follow (trajectory objects are constructed from [PathWeaver ](https://docs.wpilib.org/en/stable/docs/software/pathplanning/pathweaver/introduction.html)trajectory files)
- A `RamseteController`, which sort of acts like a PID controller but for trajectory following. A high level description can be found [here](https://docs.wpilib.org/en/stable/docs/software/advanced-controls/trajectories/ramsete.html), and a very detailed description can be found in [this book](https://file.tavsys.net/control/controls-engineering-in-frc.pdf) starting on page 110 (do ctrl + f "ramsete" and it should be the first result). 
- PID and [feedforward ](https://docs.wpilib.org/en/stable/docs/software/advanced-controls/controllers/feedforward.html)controller objects for controlling the robot's speed
- The robot's current position and a way to output voltage to the drive wheels

The `Execute` function uses the `RamseteController` to determine the robot's goal velocity for following the trajectory. It then uses the PID and feedforward controllers to determine what voltage should be output to the motors so that the robot will achieve the goal velocity. Finally, it outputs the calculated voltage to the appropriate motors.

**Comment by gaborszita** [comment]
Put this in `Constants.h` include section to resolve the conflicts:

```
#include <math.h>
#include <units/length.h>
#include <units/velocity.h>
#include <units/acceleration.h>
#include <units/voltage.h>
#include "string"
#include "frc/I2C.h"
#include <frc/Filesystem.h>
```

But please test if it compiles before doing it though, I didn't test it. I just wanted to tell what to do cause I was the one optimizing the units library includes.
